src/Entity/Enemy.cpp

```diff
@@ -47,76 +67,187 @@ void InitEnemy(flecs::world& ecs)
+		return;
+	//Reset time only when spawn was succesfull?
 	lastSpawnTime = currentTime;
+	// Spawn the enemy
+	SpawnEnemy(iter.world(), cornerIndex, lastSpawnTime, currentTime);
 
-	const TransformComponent transformComponent{{0.0f, 0.0f}, gCharacterSize};
-	const Rectangle newGoblinRect{transformComponent.mPosition.x, transformComponent.mPosition.y, gCharacterSize.x, gCharacterSize.y};
+}
 
-	bool hasCollided{false};
+bool CheckAndStartNewWave(float& lastWaveEndTime, int& cornerIndex, float currentTime)
+{
+	// Determine the current state: wave or pause
+	bool inWave = (currentTime - lastWaveEndTime) < (WAVE_DURATION + PAUSE_DURATION) && (currentTime - lastWaveEndTime) > PAUSE_DURATION;
+
+	if (!inWave)
+	{
+		// We're in a pause
+		if (currentTime - lastWaveEndTime > PAUSE_DURATION)
+		{
+			// The pause is over, start a new wave
+			lastWaveEndTime = currentTime + WAVE_DURATION;
+			cornerIndex = GetRandomValue(0, 3);
+			inWave = true;
+		}
+	}
+
+	//DEBUG:
+	gWaveInProgressDebug = inWave? 1.0f : 0.0f;
+	gCurrentWaveDurationDebug = currentTime - lastWaveEndTime;
 
-	//Save world for efficiency
-	flecs::world ecs = iter.world();
 	
-	auto filter = ecs.filter<CollisionComponent, TransformComponent>();
+	return inWave;
+}
 
-	filter.iter([&hasCollided, &newGoblinRect](flecs::iter& it, CollisionComponent* collisionComponents, TransformComponent* transformComponents) {
-		if (hasCollided)
-			return;
-		for (int i = 0; i < it.count(); ++i)
-		{
-			auto otherCollisionComponent = collisionComponents[i];
-			auto otherTransformComponent = transformComponents[i];
-			const Rectangle otherRect{otherTransformComponent.mPosition.x, otherTransformComponent.mPosition.y, otherCollisionComponent.mRectScale.x, otherCollisionComponent.mRectScale.y};
+void SpawnEnemy(flecs::world& ecs, int cornerIndex, float& lastSpawnTime, float currentTime)
+{
+	Vector2 spawnPosition{};
+	
+	//Picking map bounds to spawn from the corners of the map
+	const Texture2D mapTexture = ecs.component<MapTag>().target<MapTag>().get_ref<TextureComponent>()->mTexture;
 
-			if (CheckCollisionRecs(newGoblinRect, otherRect))
-			{
-				hasCollided = true;
-				return;
-			}
-		}
-	});
+	switch (cornerIndex)
+	{
+		case 0:
+			spawnPosition = {0.0f + gCharacterSize.x, 0.0f + gCharacterSize.y};
+			break;
+		case 1:
+			spawnPosition = {0.0f + gCharacterSize.x, (float)mapTexture.height - gCharacterSize.y};
+			break;
+		case 2:
+			spawnPosition = {(float)mapTexture.width - gCharacterSize.x, 0.0f + gCharacterSize.y};
+			break;
+		default:
+			spawnPosition = {(float)mapTexture.width - gCharacterSize.x, (float)mapTexture.height - gCharacterSize.y};
+			break;
+	}
+	
+	const TransformComponent transformComponent{spawnPosition, gCharacterSize};
+	const Rectangle newGoblinRect{transformComponent.mPosition.x, transformComponent.mPosition.y, gCharacterSize.x, gCharacterSize.y};
 
+	//Picking a tower
+	TargetTowerComponent targetTowerComponent = PickTowerTarget(ecs, transformComponent);
```

Comment:
We can move this call directly in `.set<TowerTargetComponent>({})`, because if we call it now and there’s a collision, then we did extra computations